File: grosse-beispiele/beispiel.py
from abc import ABC, abstractmethod

# Abstrakte Klasse Animal
# Definiert grundlegende Eigenschaften und abstrakte Methoden, die von Unterklassen implementiert werden müssen.
class Animal(ABC):
    def __init__(self, name, age):
        self._name = name
        self._age = age

    # Eigenschafts-Getter zur Kapselung
    @property
    def name(self):
        return self._name

    @property
    def age(self):
        return self._age

    # Abstrakte Methode, die in jeder Tier-Unterklasse definiert werden muss
    @abstractmethod
    def make_sound(self):
        pass

# Interface Feedable
# Dieses Interface stellt sicher, dass fütterbare Tiere eine "feed"-Methode haben.
class Feedable:
    def feed(self):
        raise NotImplementedError

# Klasse Lion, die von Animal erbt und Feedable implementiert
class Lion(Animal, Feedable):
    def make_sound(self):
        print(f"{self.name} roars!")

    def feed(self):
        print(f"{self.name} is being fed with meat.")

# Klasse Elephant, die von Animal erbt und Feedable implementiert
class Elephant(Animal, Feedable):
    def make_sound(self):
        print(f"{self.name} trumpets!")

    def feed(self):
        print(f"{self.name} is being fed with vegetables.")

# ZooKeeper-Klasse repräsentiert die Assoziation zwischen Pfleger und Tier
# Der ZooKeeper kann Tiere füttern, die er aber nicht besitzt.
class ZooKeeper:
    def __init__(self, name):
        self.name = name

    def care_for_animal(self, animal):
        animal.feed()

# Klasse Zoo repräsentiert den Zoo als Ganzes (Aggregation)
# Der Zoo besitzt eine Liste von Tieren, aber die Tiere existieren unabhängig vom Zoo.
class Zoo:
    def __init__(self, keeper):
        self.keeper = keeper
        self.animals = []

    def add_animal(self, animal):
        self.animals.append(animal)

    def make_all_sounds(self):
        for animal in self.animals:
            animal.make_sound()

    def feed_all_animals(self):
        for animal in self.animals:
            if isinstance(animal, Feedable):
                self.keeper.care_for_animal(animal)

# Hauptprogramm zur Verwaltung des Zoos
keeper = ZooKeeper("John")
zoo = Zoo(keeper)

lion = Lion("Simba", 5)
elephant = Elephant("Dumbo", 10)

zoo.add_animal(lion)
zoo.add_animal(elephant)

zoo.make_all_sounds()
zoo.feed_all_animals()

File: grosse-beispiele/beispiel.java
import java.util.ArrayList;
import java.util.List;

// Abstrakte Klasse Animal
// Diese Klasse definiert gemeinsame Eigenschaften für alle Tiere und enthält abstrakte Methoden,
// die von jeder Tierunterklasse überschrieben werden müssen.
abstract class Animal {
    private String name;
    private int age;

    public Animal(String name, int age) {
        this.name = name;
        this.age = age;
    }

    // Getter-Methoden, die die Kapselung der Eigenschaften "name" und "age" sicherstellen
    public String getName() {
        return name;
    }

    public int getAge() {
        return age;
    }

    // Abstrakte Methode, die von jeder Unterklasse definiert werden muss
    public abstract void makeSound();
}

// Interface Feedable
// Dieses Interface legt fest, dass jedes fütterbare Tier eine feed-Methode haben muss.
interface Feedable {
    void feed();
}

// Klasse Lion, die von Animal erbt und Feedable implementiert
// Lion erbt die Eigenschaften und Methoden von Animal und implementiert das Feedable-Interface.
class Lion extends Animal implements Feedable {
    public Lion(String name, int age) {
        super(name, age);
    }

    @Override
    public void makeSound() {
        System.out.println(getName() + " roars!");
    }

    @Override
    public void feed() {
        System.out.println(getName() + " is being fed with meat.");
    }
}

// Klasse Elephant, die von Animal erbt und Feedable implementiert
class Elephant extends Animal implements Feedable {
    public Elephant(String name, int age) {
        super(name, age);
    }

    @Override
    public void makeSound() {
        System.out.println(getName() + " trumpets!");
    }

    @Override
    public void feed() {
        System.out.println(getName() + " is being fed with vegetables.");
    }
}

// Klasse ZooKeeper, die die Beziehung zwischen Pfleger und Tier (Assoziation) repräsentiert
// Der ZooKeeper kann Tiere füttern, besitzt diese aber nicht.
class ZooKeeper {
    private String name;

    public ZooKeeper(String name) {
        this.name = name;
    }

    public void careForAnimal(Feedable animal) {
        animal.feed();
    }
}

// Klasse Zoo repräsentiert den Zoo als Ganzes (Aggregation)
// Der Zoo besitzt eine Liste von Tieren, die aber unabhängig existieren können.
class Zoo {
    private List<Animal> animals;
    private ZooKeeper keeper;

    public Zoo(ZooKeeper keeper) {
        this.keeper = keeper;
        this.animals = new ArrayList<>();
    }

    public void addAnimal(Animal animal) {
        animals.add(animal);
    }

    public void makeAllSounds() {
        for (Animal animal : animals) {
            animal.makeSound();
        }
    }

    public void feedAllAnimals() {
        for (Animal animal : animals) {
            if (animal instanceof Feedable) {
                keeper.careForAnimal((Feedable) animal);
            }
        }
    }
}

// Hauptklasse zur Ausführung der Zoo-Management-Funktionalität
public class ZooManagement {
    public static void main(String[] args) {
        ZooKeeper keeper = new ZooKeeper("John");
        Zoo zoo = new Zoo(keeper);

        Lion lion = new Lion("Simba", 5);
        Elephant elephant = new Elephant("Dumbo", 10);

        zoo.addAnimal(lion);
        zoo.addAnimal(elephant);

        zoo.makeAllSounds();
        zoo.feedAllAnimals();
    }
}
